gax/src/main/java/com/google/api/gax/retrying/BasicRetryingFuture.java

```diff
@@ -0,0 +1,216 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.api.gax.retrying;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.common.util.concurrent.AbstractFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * For internal use only.
+ *
+ * <p>Basic implementation of {@link RetryingFuture} interface. Suitable for usage in "busy loop"
+ * retry implementations.
+ *
+ * <p>This class is thread-safe.
+ */
+class BasicRetryingFuture<ResponseT> extends AbstractFuture<ResponseT>
+    implements RetryingFuture<ResponseT> {
+
+  final Object lock = new Object();
+
+  private final Callable<ResponseT> callable;
+
+  private final RetryAlgorithm<ResponseT> retryAlgorithm;
+
+  private volatile TimedAttemptSettings attemptSettings;
+
+  private volatile ApiFuture<ResponseT> latestCompletedAttemptResult;
+  private volatile ApiFuture<ResponseT> attemptResult;
+
+  BasicRetryingFuture(Callable<ResponseT> callable, RetryAlgorithm<ResponseT> retryAlgorithm) {
+    this.callable = checkNotNull(callable);
+    this.retryAlgorithm = checkNotNull(retryAlgorithm);
+
+    this.attemptSettings = retryAlgorithm.createFirstAttempt();
+
+    // A micro crime, letting "this" reference to escape from constructor before initialization is
+    // completed (via internal non-static class CompletionListener). But it is guaranteed to be ok,
+    // since listener is guaranteed to be called only after this future is
+    // completed and this future is guaranteed to be completed only after it is initialized. Also
+    // since "super" is called explicitly here there are no unexpected overrides of addListener here.
+    super.addListener(new CompletionListener(), MoreExecutors.directExecutor());
+  }
+
+  @Override
+  public void setAttemptFuture(ApiFuture<ResponseT> attemptFuture) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  // "super." is used here to avoid infinite loops of callback chains
+  public void setAttempt(Throwable throwable, ResponseT response) {
```

Comment:
I did not call it setAttemptResult becaus it will interfere with getAttemptResult which, I thought, was a different thing. But now, when I think about it, this method is literally a legit setAttemptResult method, and getAttemptResult/setAttemptResult make a valid setter/getter pair =). So yes, I'll rename it.